storage/src/mmr/verification.rs

```diff
@@ -90,6 +90,58 @@ impl Proof {
         }
         *root_hash == mmr_hasher.root_hash(self.size, peak_hashes.iter())
     }
+
+    /// Return the maximum size in bytes of any serialized `Proof`.
+    pub fn max_serialization_size<H: CHasher>() -> usize {
+        9 + (u8::MAX as usize * H::len())
+    }
+
+    /// Canonically serializes the `Proof` as:
+    /// ```text
+    ///    [0-8): size (u64 big-endian)
+    ///    [9-10): number of hashes (u8)
+    ///    [11-...): raw bytes of each hash, each of length `H::len()`
+    /// ```
+    pub fn serialize<H: CHasher>(&self) -> Vec<u8> {
+        let bytes_len = 8 + 1 + (self.hashes.len() * H::len());
+        let mut bytes = Vec::with_capacity(bytes_len);
+        bytes.extend_from_slice(&self.size.to_be_bytes());
+
+        // A proof should never contain more hashes than the depth of the MMR, thus a single byte
+        // for encoding the length of the hashes array still allows serializing MMRs up to 2^255
+        // elements.
+        assert!(
+            self.hashes.len() <= u8::MAX as usize,
+            "too many hashes in proof"
+        );
+        bytes.push(self.hashes.len() as u8);
+        for hash in self.hashes.iter() {
+            bytes.extend_from_slice(hash);
+        }
+        if bytes.len() != bytes_len {
+            panic!("serialization length mismatch");
+        }
+        bytes.to_vec()
+    }
+
+    /// Deserializes a canonically encoded `Proof`. See `serialize` for the serialization format.
+    pub fn deserialize<H: CHasher>(bytes: &[u8]) -> Option<Self> {
+        if bytes.len() < 9 {
+            return None;
+        }
+        let size = u64::from_be_bytes(bytes[0..8].try_into().unwrap());
+        let hashes_len = bytes[8] as usize;
+        let mut hashes = Vec::with_capacity(hashes_len);
+        if bytes.len() != 9 + hashes_len * H::len() {
```

Comment:
Might be worth commenting why checked addition isn't necessary? (H::len() although a usize should never be close to bounds and the byte is up to 255).

Reply:
Other approach: just do checked ops lol.

Reply:
Hmm, checked math is pretty ugly. I just added a comment if that's OK.